### DISCUSSION TITLE
`repeat` parsers must always consume

### DISCUSSION THREAD

**Discussion by woile** [winnow-rs/winnow#508] [Ideas]
Hi there, I'm opening this thread to discuss a bit further how to improve `repeat` egonomics as exchanged with [epage on mastodon](https://hachyderm.io/@woile/112275678303839449).

Some context first, I [migrated recipe-lang](https://github.com/reciperium/recipe-lang/commit/80a6a1d70bd4840489fbb068bfbcbe417f54c971) from `nom` to `winnow`. recipe-lang, parses a recipe as a `Vec<Token>`, as opposed to a json-parser, where a single data structure is received, and you can recursively parse "internal" data structures if there's any (for example and object may contain more objects).

In order to parse the `Vec<Token>` with nom, I was using [many1](https://github.com/reciperium/recipe-lang/blob/v0.3.1/crates/recipe-parser/src/parser.rs#L238), when I switched to `winnow` I tried using `repeat` which kept throwing an error:

> `repeat` parsers must always consume

Eventually, I figured I could use [repeat_till](https://docs.rs/winnow/latest/winnow/combinator/fn.repeat_till.html) until the `eof`. That worked. 

Some things maybe to discuss:

1. Can we improve the `repeat` error? how to solve it, or why it happens?. I couldn't figure it out. 
2. I was searching by the alias `many1` and only `repeat` is listed, not `repeat_till`, should we add an alias to repeat_till?
3. Was my approach correct? is there a way I can still use `repeat`? 

Regards

**Comment by epage** [comment]
> In order to parse the Vec<Token> with nom, I was using [many1](https://github.com/reciperium/recipe-lang/blob/v0.3.1/crates/recipe-parser/src/parser.rs#L238), when I switched to winnow I tried using repeat which kept throwing an error:

nom also errored in these cases (see [multi/mod.rs](https://docs.rs/nom/latest/src/nom/multi/mod.rs.html#128-131)), we just changed that from an error that can be backtracked to an assert.  We did that with the assumption that this scenario is a programmer error and should be resolved but it could be masked, depending on the design of the parser.

What would be helpful to know is why you hit that situation so we can explore what the right path is, whether in recipe-lang, winnow, or both.  Looking at the code, it isn't immediately obvious.  Could you step through how the `repeat`ed parser was consuming none of the input?

The reason this error hits is to prevent infinite loops.  If you do `repeat(0.., empty)` without this check, you'll get an infinite loop.  Now, that is a more obvious, discoverable cases.  Whats harder is when a parser sometimes consumes none of the input.

> Can we improve the repeat error? how to solve it, or why it happens?. I couldn't figure it out.

Thats reasonable.

Currently the error is "`repeat` parsers must always consume" and there is a code-comment "infinite loop check: the parser must always consume".  We likely should combine those two

```
`repeat`ed parser did not consume any of the input which it must to prevent `repeat`ing indefinitely
```

> I was searching by the alias many1 and only repeat is listed, not repeat_till, should we add an alias to repeat_till?

`repeat_till` is a very different parser.  I'm not too sure why we would add a `many1` alias to it.  We do have a `many_till0` alias on it because that is the name of the parser in `nom`.

> Was my approach correct? is there a way I can still use repeat?

`repeat_till` seems like it should be giving you the same error.  Since it isn't, I'm assuming you could solve the problem with `repeat` as well.  We just need to better understand why it was happening.

**Reply by woile** [reply]
> nom also errored in these cases (see [multi/mod.rs](https://docs.rs/nom/latest/src/nom/multi/mod.rs.html#128-131)), we just changed that from an error that can be backtracked to an assert

That's strange, the link I shared was the old version of recipe-lang using many1, and it was not throwing any errors:
https://github.com/reciperium/recipe-lang/blob/v0.3.1/crates/recipe-parser/src/parser.rs#L238

That's why I was surprised, I migrated the nom code to winnow and besides changing `&str` to `&mut str` and using `parse_next` and `parse`, everything else was almost 1 on 1 swap.

Here's where the `repeat_till` is used:
https://github.com/reciperium/recipe-lang/blob/v0.4.2/crates/recipe-parser/src/parser.rs#L300

When I change it for `repeat` the tests start failing:
https://github.com/reciperium/recipe-lang/blob/v0.4.2/crates/recipe-parser/src/parser.rs#L553

I'm gonna give it another try with the repeat

**Reply by baoyachi** [reply]
@woile The changes in the core code are as follows:

Associated code link:https://github.com/reciperium/recipe-lang/commit/80a6a1d70bd4840489fbb068bfbcbe417f54c971


```diff 
pub fn recipe_value<'a>(input: &mut Input<'a>) -> PResult<Token<'a>> {
    alt((
        parse_metadata.map(|(key, value)| Token::Metadata { key, value }),
        parse_material.map(|m| Token::Material(m)),
        parse_timer.map(|t| Token::Timer(t)),
        // Because ingredient doesn't have a prefix before the curly braces, e.g: `&{}`
        // it must always be parsed after timer and material
        parse_ingredient.map(|(name, amount)| {
            let mut quantity = None;
            let mut unit = None;
            if let Some((_quantity, _unit)) = amount {
                quantity = _quantity;
                unit = _unit;
            };

            Token::Ingredient {
                name,
                quantity,
                unit,
            }
        }),
        parse_recipe_ref.map(|(name, amount)| {
            let mut quantity = None;
            let mut unit = None;
            if let Some((_quantity, _unit)) = amount {
                quantity = _quantity;
                unit = _unit;
            };

            Token::RecipeRef {
                name,
                quantity,
                unit,
            }
        }),
        parse_backstory.map(|v| Token::Backstory(v)),
        parse_comment.map(|v| Token::Comment(v)),
        "(".map(|v| Token::Word(v)),
        parse_word.map(|v| Token::Word(v)),
+        multispace1.map(|v| Token::Space(v)),
+        // rest.map(|v| Token::Word(v)),
+        // eof.map(|_| Token::Space("")),
    ))
    .parse_next(input)
}

pub fn recipe<'a>(input: &mut Input<'a>) -> PResult<Vec<Token<'a>>> {
    // repeat_till(0.., recipe_value, eof).parse_next(input)
+    repeat(0..,recipe_value).parse_next(input)
}
```

Based on the code provided above, all `recipe_parser` crates unit tests can now be run.

**Reply by epage** [reply]
> That's strange, the link I shared was the old version of recipe-lang using many1, and it was not throwing any errors:
https://github.com/reciperium/recipe-lang/blob/v0.3.1/crates/recipe-parser/src/parser.rs#L238

That doesn't have a `rest` or `eof` in it but in the winnow versions I tend to see those present but commented out.  Why is it?

**Reply by woile** [reply]
> That doesn't have a rest or eof in it but in the winnow versions I tend to see those present but commented out. Why is it?

oh they are not needed, I added them to test if the `repeat` would work. With or without them, the parser fails on `repeat`

I tried @baoyachi implementation and it works, using `multispace1`. Apparently my `parse_space` implementation gets into a conflict with `repeat`:

```rs
fn parse_space<'a>(input: &mut Input<'a>) -> PResult<&'a str> {
    let multispace = " \t\r\n";
    take_while(0.., move |c| multispace.contains(c)).parse_next(input)
}
```

And, if I modify it to:

```diff
fn parse_space<'a>(input: &mut Input<'a>) -> PResult<&'a str> {
    let multispace = " \t\r\n";
-    take_while(0.., move |c| multispace.contains(c)).parse_next(input)
+    take_while(1.., move |c| multispace.contains(c)).parse_next(input)
}
```

everything works

**Reply by woile** [reply]
Thanks for thinking along, this is the PR with the changes: https://github.com/reciperium/recipe-lang/pull/4
I went with `multispace1` so I don't have to reimplement it

**Reply by baoyachi** [reply]
yeah. The `alt` combinator in Winnow requires each branch to maintain determinism, either successful or reporting an error (such as backtrack error). Previously, `take_while(0..)` was used, which allows for parsing even when the input is empty, resulting in errors in the internal repeat_ method of Winnow. For details, please refer to:https://github.com/winnow-rs/winnow/blob/73211ad0e7c830cc926fc52f1818425fb8918040/src/combinator/multi.rs#L322-L324.


Currently, if you set take_while (1..), when the input is empty, the method will report a backtrack error, and alt will attempt the next combinator.

**Reply by baoyachi** [reply]
For the difference in the `Into<Range>` parameter of `take_while`, please refer to the documentation for usage examples: https://docs.rs/winnow/latest/winnow/token/fn.take_while.html

**Reply by woile** [reply]
I run a simple benchmark and the winnow version is now slightly faster 🚀  (on a mac, and a file with 1k lines)

```shell
❯ hyperfine --shell=none --warmup 1 './recp-nom show long-hummus.recp' './recp-win show long-hummus.recp'
Benchmark 1: ./recp-nom show long-hummus.recp
  Time (mean ± σ):       9.2 ms ±   0.7 ms    [User: 5.4 ms, System: 2.0 ms]
  Range (min … max):     8.1 ms …  14.4 ms    246 runs
 
Benchmark 2: ./recp-win show long-hummus.recp
  Time (mean ± σ):       9.1 ms ±   0.7 ms    [User: 5.4 ms, System: 1.9 ms]
  Range (min … max):     8.1 ms …  13.5 ms    307 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  ./recp-win show long-hummus.recp ran
    1.01 ± 0.11 times faster than ./recp-nom show long-hummus.recp
```

**Reply by epage** [reply]
What still confuses me is that in nom, `parse_space` should have consumed nothing and then `many1` should have seen that and errored.  As `repeat` is at the top of the stack, that should have been reported to the user, rather than backteacked.
